keps/sig-api-machinery/20200309-consistent-resource-versions-semantics/README.md

```diff
@@ -0,0 +1,236 @@
+from the code setting a limit on the request. For example, ListerWatchers can be
+layered in to add pagination limits.
+
+This was discussed in [Feb 12th, 2020 SIG API machinery bi-weekly meeting](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#bookmark=id.3kvpricxohe8).
+
+### Goals
+
+- Improve API to make resource version semantics consistent.
+- Backward compaibility. In particular, support the case of a client making a request using the
+  updated API against an server serving the old API.
+
+## Proposal
+
+### Add a ResourceVersionMatch query parameter
+
+Add an optional `ResourceVersionMatch` paramater to `ListOptions` and
+`GetOptions` with the enumeration values:
+
+```
+// ResourceVersionMatch specifies how the ResourceVersion parameter is applied. ResourceVersionMatch
+// only applies to list calls and may only be set if ResourceVersion is also set.
+type ResourceVersionMatch string
+
+const (
+	// ResourceVersionMatchNotOlderThan matches data at least as new as the provided
+	// ResourceVersion. The newest available data is preferred, but any data not
+	// older than this ResourceVersion may be served.
+	// For list calls, this guarantees that ResourceVersion in the ListMeta is not older than the requested
+	// ResourceVersion, but does not make any guarantee about the ResourceVersion in the ObjectMeta
+	// of the list items since ObjectMeta.ResourceVersion tracks when an object was last updated,
+	// not how up-to-date the object is when served.
+	ResourceVersionMatchNotOlderThan ResourceVersionMatch = "NotOlderThan"
+	// ResourceVersionMatchExact matches data at the exact ResourceVersion
+	// provided. If the provided ResourceVersion is unavailable, the server responds with
+	// HTTP 410 “Gone”.
+	// For list calls, this guarantees that ResourceVersion in the ListMeta is the same as the requested
+	// ResourceVersion, but does not make any guarantee about the ResourceVersion in the ObjectMeta
+	// of the list items since ObjectMeta.ResourceVersion tracks when an object was last updated,
+	// not how up-to-date the object is when served.
+	ResourceVersionMatchExact ResourceVersionMatch = "Exact"
+)
+```
+
+```
+type ListOptions struct {
+    ...
+	// ResourceVersion sets a constraint on what resource versions a request may be served from.
+	// If unset, the request is served from latest resource version to ensure strong consistency
+	// (i.e. served from etcd via a quorum read). Setting a ResourceVersion is preferable in cases
+	// where a ResourceVersion is known and can be used as a constraint since better performance
+	// and scalability can be achieved for a cluster by avoiding quorum reads.
+	// When ResourceVersion is set for list, it is highly recommended ResourceVersionMatch is also set.
+	//
+	// ResourceVersion for watch:
+	// - if unset, start a watch at the most recent resource version, which must be consistent
```

Comment:
I'll clean this up.

Reply:
Okay, hadn't documented any consistency guarantee before I expanded this documentation, so we can safely clarify it without contradicting any pre-existing statements.

OK, since watches from storage do happen to provide a [consistency guarantee](https://github.com/kubernetes/kubernetes/blob/c158001bbc77b8bf13d8761831cce79e3b6fedec/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L183) or RV=0, but the watch cache doesn't, we should document this as not providing a guarantee. This means that we don't distinguish between RV="0" and RV="" for watch (and the code in both code paths represents resource versions at golang ints where RV="" is converted to 0). I'll consolidate the documentation for those two cases. @wojtek-t let me know if that doesn't sound right to you.